schemaregistry/rest_service.go

```diff
@@ -0,0 +1,293 @@
+			return err
+		}
+	}
+
+	tlsConf.BuildNameToCertificate()
+
+	return err
+}
+
+// configureTransport returns a new Transport for use by the Confluent Schema Registry REST client
+func configureTransport(conf *Config) (*http.Transport, error) {
+
+	// Exposed for testing purposes only. In production properly formed certificates should be used
+	// https://tools.ietf.org/html/rfc2818#section-3
+	tlsConfig := &tls.Config{}
+	if err := configureTLS(conf, tlsConfig); err != nil {
+		return nil, err
+	}
+
+	timeout := conf.ConnectionTimeoutMs
+
+	return &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: time.Duration(timeout) * time.Millisecond,
+		}).Dial,
+		TLSClientConfig: tlsConfig,
+	}, nil
+}
+
+// configureURLAuth copies the url userinfo into a basic HTTP auth authorization header
+func configureURLAuth(service *url.URL, header http.Header) error {
+	header.Add("Authorization", fmt.Sprintf("Basic %s", encodeBasicAuth(service.User.String())))
+	return nil
+}
+
+// configureSASLAuth copies the sasl username and password into a HTTP basic authorization header
+func configureSASLAuth(conf *Config, header http.Header) error {
+	mech := conf.SaslMechanism
+	if strings.ToUpper(mech) == "GSSAPI" {
+		return fmt.Errorf("SASL_INHERIT support PLAIN and SCRAM SASL mechanisms only")
+	}
+
+	user := conf.SaslUsername
+	pass := conf.SaslPassword
+	if user == "" || pass == "" {
+		return fmt.Errorf("SASL_INHERIT requires both sasl.username and sasl.password be set")
+	}
+
+	header.Add("Authorization", fmt.Sprintf("Basic %s", encodeBasicAuth(fmt.Sprintf("%s:%s", user, pass))))
+	return nil
+}
+
+// configureUSERINFOAuth copies basic.auth.user.info
+func configureUSERINFOAuth(conf *Config, header http.Header) error {
+	auth := conf.BasicAuthUserInfo
+	if auth == "" {
+		return fmt.Errorf("USER_INFO source configured without basic.auth.user.info ")
+	}
+
+	header.Add("Authorization", fmt.Sprintf("Basic %s", encodeBasicAuth(auth)))
+	return nil
+
+}
+
+// newAuthHeader returns a base64 encoded userinfo string identified on the configured credentials source
+func newAuthHeader(service *url.URL, conf *Config) (http.Header, error) {
+	// Remove userinfo from url regardless of source to avoid confusion/conflicts
+	defer func() {
+		service.User = nil
+	}()
+
+	source := conf.BasicAuthCredentialsSource
+
+	header := http.Header{}
+
+	var err error
+	switch strings.ToUpper(source) {
+	case "URL":
+		err = configureURLAuth(service, header)
+	case "SASL_INHERIT":
+		err = configureSASLAuth(conf, header)
+	case "USER_INFO":
+		err = configureUSERINFOAuth(conf, header)
+	default:
+		err = fmt.Errorf("unrecognized value for basic.auth.credentials.source %s", source)
+	}
+	return header, err
+}
+
+// handleRequest sends a HTTP(S) request to the Schema Registry, placing results into the response object
+func (rs *restService) handleRequest(request *api, response interface{}) error {
+	endpoint, err := rs.url.Parse(fmt.Sprintf(base+request.endpoint, request.arguments...))
```

Comment:
Hi @emasab, @edenhill,

Curious why is this done?  This seems to break for schema registry servers that are located at a specific path, not just a hostname.  For example if my schema registry endpoint is at "https://foo.bar/baz", and I want to call the GET "subjects" API, then this will send the request to host https://foo.bar with path "/subjects" instead of "/baz/subjects".

Was this an attempt to trim the potentially-trailing slash?  I.e. to change "https://foo.bar/baz/" to "https://foo.bar/baz"?